src/doc-utils.js

```diff
@@ -41,26 +51,92 @@ function isEmpty(n) {
 }
 
 function getFirstString(doc) {
-  return iterDoc(doc, (type, doc) => {
-    if (type === "string" && doc.trim().length !== 0) {
-      return doc;
+  let firstString = null;
+  traverseDoc(doc, doc => {
+    if (
+      typeof doc === "string" && doc.trim().length !== 0 && firstString === null
+    ) {
+      firstString = doc;
+    }
+  });
+  return firstString;
+}
+
+function isLineNext(doc) {
+  let flag = null;
+  traverseDoc(doc, doc => {
+    if(flag === null) {
+      if (typeof doc === "string") {
+        flag = false;
+      }
+      if (doc.type === "line") {
+        flag = true;
+      }
     }
   });
+  return !!flag;
 }
 
-function hasHardLine(doc) {
-  // TODO: If we hit a group, check if it's already marked as a
-  // multiline group because they should be marked bottom-up.
-  return !!iterDoc(doc, (type, doc) => {
-    switch (type) {
+function willBreak(doc) {
+  let willBreak = false;
+  traverseDoc(doc, doc => {
+    switch (doc.type) {
+      case "group":
+        if(doc.break) {
+          willBreak = true;
+        }
       case "line":
         if (doc.hard) {
-          return true;
+          willBreak = true;
         }
 
         break;
     }
   });
+  return willBreak;
+}
+
+function breakParentGroup(groupStack) {
+  if(groupStack.length > 0) {
+    const parentGroup = groupStack[groupStack.length - 1];
+    // Breaks are not propagated through conditional groups because
+    // the user is expected to manually handle what breaks.
+    if(!parentGroup.expandedStates) {
+      parentGroup.break = true;
+    }
+  }
+  return null;
+}
+
+function propagateBreaks(doc) {
+  const groupStack = [];
+  traverseDocUpDown(
+    doc,
+    doc => {
+      if (doc.type === "break-parent") {
+        breakParentGroup(groupStack);
+      }
+      if (doc.type === "group") {
+        groupStack.push(doc);
+      }
+    },
+    doc => {
+      if (doc.type === "group") {
+        const group = groupStack.pop();
+        if(group.break) {
+          breakParentGroup(groupStack);
```

Comment:
> because otherwise you'll be iterating a lot over groups that have already been marked

If a group has already been marked, you can just stop the traversal. You don't need to keep traversing the same nodes over and over again.



Tangential but you may be interested in this post I wrote: http://blog.vjeux.com/2015/javascript/dirty-up-and-execute-top-down.html